packages/gatsby/src/utils/start-server.ts

```diff
@@ -213,6 +216,51 @@ export async function startServer(
     res.end()
   })
 
+  app.get(
+    `/page-data/:pagePath(*)/page-data.json`,
+    async (req, res, next): Promise<void> => {
+      const requestedPagePath = req.params.pagePath
+      if (!requestedPagePath) {
+        report.verbose(`[page-data-handler] empty requestedPagePath, skipping.`)
+        next()
+        return
+      }
+
+      const potentialPagePath = reverseFixedPagePath(requestedPagePath)
+      const page = findPageByPath(store.getState(), potentialPagePath, false)
+
+      if (page) {
+        report.verbose(
+          `[page-data-handler] page for "${requestedPagePath}":\n${JSON.stringify(
+            page,
+            null,
+            2
+          )}`
+        )
+
+        try {
+          const pageData = await readPageData(
+            path.join(store.getState().program.directory, `public`),
+            page.path
+          )
```

Comment:
Just to clarify - we will actually use the new page data utility here? (i.e. #27939)

Reply:
That's the plan. It should look similar as https://github.com/gatsbyjs/gatsby/pull/27939/files#diff-e4535fc084e00493cdf64059b0257b3ba2ee3fe54ec9396daf87158b0679905cL193-R200 part in websocket manager